#!/usr/bin/env python3
"""
üîß Test de Conexi√≥n ZeroMQ 5559
Verifica que el Enhanced Promiscuous Agent est√© enviando eventos
"""

import zmq
import time
import logging
from datetime import datetime

# Configurar logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Intentar importar protobuf
PROTOBUF_AVAILABLE = False
try:
    from src.protocols.protobuf import network_event_pb2

    PROTOBUF_AVAILABLE = True
    logger.info("‚úÖ Protobuf disponible: src.protocols.protobuf.network_event_pb2")
except ImportError:
    try:
        import network_event_pb2

        PROTOBUF_AVAILABLE = True
        logger.info("‚úÖ Protobuf disponible: directorio local")
    except ImportError:
        logger.warning("‚ö†Ô∏è Protobuf NO disponible - eventos se mostrar√°n como raw bytes")


def test_zmq_connection():
    """Test de conexi√≥n a ZeroMQ 5559"""
    print("üîß TEST DE CONEXI√ìN ZeroMQ 5559")
    print("=" * 40)
    print(f"üì¶ Protobuf disponible: {PROTOBUF_AVAILABLE}")
    print("üéØ Conectando a tcp://localhost:5559...")
    print("‚è±Ô∏è Esperando eventos por 30 segundos...")
    print("üõë Presiona Ctrl+C para detener")
    print("")

    context = None
    socket = None
    events_received = 0
    events_with_gps = 0

    try:
        # Configurar ZeroMQ
        context = zmq.Context()
        socket = context.socket(zmq.SUB)
        socket.connect("tcp://localhost:5559")
        socket.setsockopt(zmq.SUBSCRIBE, b"")
        socket.setsockopt(zmq.RCVTIMEO, 1000)  # Timeout 1 segundo

        print("‚úÖ Conectado a ZeroMQ puerto 5559")

        start_time = time.time()
        last_event_time = None

        while time.time() - start_time < 30:  # 30 segundos m√°ximo
            try:
                # Recibir mensaje
                message = socket.recv(zmq.NOBLOCK)
                events_received += 1
                last_event_time = datetime.now()

                if PROTOBUF_AVAILABLE:
                    # Decodificar protobuf
                    event = network_event_pb2.NetworkEvent()
                    event.ParseFromString(message)

                    has_gps = event.latitude != 0 and event.longitude != 0
                    if has_gps:
                        events_with_gps += 1

                    print(f"üì° Evento #{events_received}: {event.source_ip} ‚Üí {event.target_ip}")
                    print(f"   üéØ Puerto: {event.dest_port} | Agente: {event.agent_id}")
                    print(f"   üìä Anomal√≠a: {event.anomaly_score:.3f} | Riesgo: {event.risk_score:.3f}")
                    if has_gps:
                        print(f"   üó∫Ô∏è GPS: {event.latitude:.6f}, {event.longitude:.6f}")
                    print(f"   ‚è∞ {last_event_time.strftime('%H:%M:%S')}")
                    print("")
                else:
                    # Mostrar bytes raw
                    print(f"üì° Evento #{events_received}: {len(message)} bytes")
                    print(f"   Raw: {message[:50]}{'...' if len(message) > 50 else ''}")
                    print(f"   ‚è∞ {last_event_time.strftime('%H:%M:%S')}")
                    print("")

            except zmq.Again:
                # No hay mensajes, continuar
                time.sleep(0.1)
                continue

        # Resumen final
        print("üìä RESUMEN DEL TEST:")
        print(f"   ‚úÖ Eventos recibidos: {events_received}")
        if PROTOBUF_AVAILABLE:
            print(f"   üó∫Ô∏è Eventos con GPS: {events_with_gps}")
            print(f"   üìç Porcentaje GPS: {(events_with_gps / max(1, events_received)) * 100:.1f}%")
        print(f"   ‚è∞ √öltimo evento: {last_event_time if last_event_time else 'Ninguno'}")

        if events_received > 0:
            print("\n‚úÖ CONEXI√ìN EXITOSA - El dashboard funcionar√° correctamente")
        else:
            print("\n‚ö†Ô∏è NO SE RECIBIERON EVENTOS")
            print("   Verifica que el Enhanced Promiscuous Agent est√©:")
            print("   - Ejecut√°ndose correctamente")
            print("   - Enviando eventos al puerto ZeroMQ 5559")
            print("   - Capturando tr√°fico de red")

    except KeyboardInterrupt:
        print(f"\nüõë Test interrumpido por usuario")
        print(f"üìä Eventos recibidos hasta ahora: {events_received}")

    except Exception as e:
        print(f"\n‚ùå ERROR DE CONEXI√ìN: {e}")
        print("üîß Posibles soluciones:")
        print("   1. Verifica que el Enhanced Promiscuous Agent est√© ejecut√°ndose")
        print("   2. Confirma que usa el puerto ZeroMQ 5559")
        print("   3. Verifica que no hay firewall bloqueando localhost:5559")

    finally:
        if socket:
            socket.close()
        if context:
            context.term()


if __name__ == "__main__":
    test_zmq_connection()